platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbCompactionCoordinator.java

```diff
@@ -0,0 +1,283 @@
+
+import com.swirlds.common.config.singleton.ConfigurationHolder;
+import com.swirlds.common.threading.framework.config.ThreadConfiguration;
+import com.swirlds.merkledb.config.MerkleDbConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.io.IOException;
+import java.nio.channels.ClosedByInterruptException;
+import java.util.Optional;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This class is responsible for coordinating compaction tasks for a {@link MerkleDbDataSource}.
+ * It provides convenient API for starting compactions for each of the three storage types. Also, this class makes sure
+ * that there are no concurrent compactions for the same storage type. And finally it provides a way to stop all compactions
+ * and keep them disabled until they are explicitly enabled again.
+ * The compaction tasks are executed in a background thread pool.
+ * The number of threads in the pool is defined by {@link MerkleDbConfig#compactionThreads()} property.
+ *
+ */
+@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+class MerkleDbCompactionCoordinator {
+
+    /**
+     * Since {@code com.swirlds.platform.Browser} populates settings, and it is loaded before any
+     * application classes that might instantiate a data source, the {@link ConfigurationHolder}
+     * holder will have been configured by the time this static initializer runs.
+     */
+    private static final MerkleDbConfig config = ConfigurationHolder.getConfigData(MerkleDbConfig.class);
+
+    /**
+     * A thread pool to run compaction tasks.
+     */
+    private static final AtomicReference<ExecutorService> compactionExecutorServiceRef = new AtomicReference<>();
+
+    private static final Logger logger = LogManager.getLogger(MerkleDbCompactionCoordinator.class);
+
+    public static final String HASH_STORE_DISK_SUFFIX = "HashStoreDisk";
+    public static final String OBJECT_KEY_TO_PATH_SUFFIX = "ObjectKeyToPath";
+    public static final String PATH_TO_KEY_VALUE_SUFFIX = "PathToKeyValue";
+    private final AtomicBoolean compactionEnabled = new AtomicBoolean();
+    // we need a map of exactly three elements, one per storage
+    final ConcurrentMap<String, InterruptibleCompletableFuture> compactionFuturesByName = new ConcurrentHashMap<>(3);
+    private final MerkleDbStatisticsUpdater statisticsUpdater;
+    private final String tableName;
+    private final Optional<Compactable> objectKeyToPathOpt;
```

Comment:
Depending on the configuration, these storages may be absent.

Reply:
Sure, but how is that different from null?  The purpose of Optional is to differentiate between `not present` and `null`.

Reply:
In this particular case it could've been null and I could've replaced `someVar.ifPresent` call with `if( someVar != null)`. I just like working with `Optional`-s more because I don't have to test additional logical branch. Yes, effectively it's the same but from the code coverage perspective it's not. 

>The purpose of Optional is to differentiate between not present and null.

I don't think it's the only purpose. Otherwise, `Optional.ofNullable()` wouldn't be a thing.